keel-web/src/test/kotlin/com/netflix/spinnaker/keel/ApplicationServiceTests.kt

```diff
@@ -0,0 +1,358 @@
+            environmentName = "production",
+            artifactVersion = "fnord-1.0.0-h0.a0a0a0a",
+            type = "manual-judgement",
+            status = ConstraintStatus.OVERRIDE_PASS,
+            createdAt = clock.start,
+            judgedAt = productionDeployed.minus(Duration.ofMinutes(30)),
+            judgedBy = "lpollo@acme.com",
+            comment = "Aye!"
+          )
+        )
+      }
+
+      test("can get resource summary by application") {
+        val summaries = applicationService.getResourceSummariesFor(application)
+
+        expect {
+          that(summaries.size).isEqualTo(3)
+          that(summaries.map { it.status }.filter { it == HAPPY }.size).isEqualTo(3)
+          that(summaries.map { it.moniker?.stack }).containsExactly("test", "staging", "production")
+        }
+      }
+
+      test("can get environment summaries by application") {
+        val summaries = applicationService.getEnvironmentSummariesFor(application)
+
+        val production = summaries.find { it.name == "production" }
+        val staging = summaries.find { it.name == "staging" }
+        val test = summaries.find { it.name == "test" }
+
+        val expectedProd = EnvironmentSummary(
+          deliveryConfig.environments.find { it.name == "production" }!!,
+          setOf(ArtifactVersions(
+            artifact.name,
+            artifact.type,
+            artifact.statuses,
+            ArtifactVersionStatus(
+              current = "fnord-1.0.0-h0.a0a0a0a",
+              pending = listOf("fnord-1.0.1-h1.b1b1b1b", "fnord-1.0.2-h2.c2c2c2c", "fnord-1.0.3-h3.d3d3d3d"),
+              approved = listOf(),
+              previous = listOf(),
+              vetoed = listOf(),
+              deploying = null
+            )
+          ))
+        )
+        val expectedStage = EnvironmentSummary(
+          deliveryConfig.environments.find { it.name == "staging" }!!,
+          setOf(ArtifactVersions(
+            artifact.name,
+            artifact.type,
+            artifact.statuses,
+            ArtifactVersionStatus(
+              current = "fnord-1.0.1-h1.b1b1b1b",
+              pending = listOf("fnord-1.0.2-h2.c2c2c2c", "fnord-1.0.3-h3.d3d3d3d"),
+              approved = listOf(),
+              previous = listOf("fnord-1.0.0-h0.a0a0a0a"),
+              vetoed = listOf(),
+              deploying = null
+            )
+          ))
+        )
+        val expectedTest = EnvironmentSummary(
+          deliveryConfig.environments.find { it.name == "test" }!!,
+          setOf(ArtifactVersions(
+            artifact.name,
+            artifact.type,
+            artifact.statuses,
+            ArtifactVersionStatus(
+              current = "fnord-1.0.2-h2.c2c2c2c",
+              pending = listOf(),
+              approved = listOf("fnord-1.0.3-h3.d3d3d3d"),
+              previous = listOf("fnord-1.0.0-h0.a0a0a0a", "fnord-1.0.1-h1.b1b1b1b"),
+              vetoed = listOf(),
+              deploying = null
```

Comment:
These turned out pretty easy to read in this format! 🏆

Reply:
SO MUCH better!

Reply:
Oh, thank you very much for my part, hahaha! 🤣